src/ir/auth_logic/lowering_ast_datalog.h

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2021 The Raksha Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! This file contains the datalog IR (DLIR) which makes the translation from
+//! this authorization logic into datalog simpler.
+#ifndef SRC_IR_AUTH_LOGIC_LOWERING_AST_DATALOG_H_
+#define SRC_IR_AUTH_LOGIC_LOWERING_AST_DATALOG_H_
+
+#include "ast.h"
+
+namespace raksha::ir::auth_logic {
+
+
+  // This implements a compiler pass for converting authorization logic 
+  // (represented by ast.h) into the datalog ir (in datalog_ir.h). The only 
+  // public interface to this class is a single function "Lower" which takes
+  // as input an authorization logic program and returns a datalog program. 
+  //
+  // (Internally, this is a stateful class because it uses a counter to
+  // generate fresh names for arguments. The public interface hides this 
+  // implementation detail.)
+  class LoweringToDatalogPass {
+   public:
+
+    static DLIRProgram* Lower(const Program& auth_logic_program) {
+      LoweringToDatalogPass* pass = new LoweringToDatalogPass();
+      DLIRProgram* dlir_program = pass->ProgToDLIR(auth_logic_program);
+      delete pass;
+      return dlir_program;
+    }
+
+   private:
+    LoweringToDatalogPass()
+      : fresh_var_count_(0) {}
+
+    int fresh_var_count_;
+
+    // This returns a fresh variable name. Fresh variables are introduced in a
+    // few places during the translation.
+    std::string FreshVar() {
+      return "x__" + std::to_string(fresh_var_count_++);
+    }
+
+    // This function makes a copy of a predicate that prepends the name with
+    // another string and prepends the list of arguments with additional 
+    // arguments. This is used in a few places in the translation, for
+    // example, to translate "X says blah(args)" into "says_blah(X, args)".
+    Predicate PushOntoPred(std::string modifier,
+        std::vector<std::string> new_args, const Predicate& predicate) {
```

Comment:
It looks like `predicate` here has its contents entirely copied into the new `Predicate`. If that is the case, it is likely better to have `Predicate` be passed as a value directly rather than as a reference (ie, `Predicate predicate` rather than `const Predicate &predicate) so that it plays nicely with copy elision when the original value can be discarded.